README.md

# iNZightPlots

![R-CMD-check](https://github.com/iNZightVIT/iNZightPlots/workflows/R-CMD-check/badge.svg)
[![codecov](https://codecov.io/gh/iNZightVIT/iNZightPlots/branch/dev/graph/badge.svg)](https://codecov.io/gh/iNZightVIT/iNZightPlots)
[![License: GPL
v3](https://img.shields.io/badge/License-GPL%20v3-blue.svg)](http://www.gnu.org/licenses/gpl-3.0)
[![CRAN](https://www.r-pkg.org/badges/version/iNZightPlots)](https://CRAN.R-project.org/package=iNZightPlots)

Graphics functions which provide a simple, easy-to-learn interface,
allowing beginners to explore and visualize data easily.

## Install

Easily install the latest release version from CRAN:

``` r
install.packages("iNZightPlots")
```

For the latest development version:

``` r
# install.packages("devtools")
devtools::install_github("iNZightVIT/iNZightTools")
devtools::install_github("iNZightVIT/iNZightMR")
devtools::install_github("iNZightVIT/iNZightPlots")
```

## Usage

`iNZightPlots` provides a simple interface to graphics, which are most
easily produced using the following function:

``` r
library(iNZightPlots)
inzplot(~Sepal.Length, data = iris)
```

<img src="man/figures/README-demo-1.png" width="50%" />

From there, formula notation can be used to explore relationships
between variables:

``` r
inzplot(Sepal.Length ~ Species, data = iris)
inzplot(Sepal.Length ~ Sepal.Width, data = iris)
inzplot(Sepal.Length ~ Sepal.Width | Species, data = iris)
```

<img src="man/figures/README-demo-usage-2.png" width="30%" />
<img src="man/figures/README-demo-usage-3.png" width="30%" />
<img src="man/figures/README-demo-usage-4.png" width="30%" />

These functions all produce different graphs (dot plot and scatter
plot), which means the focus is on exploring data, not on which type of
graph to use. Subsetting works as expected, and can include up to two
variables (e.g., `y ~ x | g1 + g2`).

There are additionally two companion functions for summary and inference
information, which are

``` r
inzsummary(~Sepal.Length, data = iris)
#> ====================================================================================================
#>                                           iNZight Summary
#> ----------------------------------------------------------------------------------------------------
#>    Primary variable of interest: Sepal.Length (numeric)
#>                                  
#>    Total number of observations: 150
#> ====================================================================================================
#> 
#> Summary of Sepal.Length:
#> ------------------------
#> 
#> Estimates
#> 
#>    Min   25%   Median   75%   Max    Mean       SD   Sample Size
#>    4.3   5.1      5.8   6.4   7.9   5.843   0.8281           150
#> 
#> 
#> ====================================================================================================
inzinference(Sepal.Length ~ Species, data = iris)
#> ====================================================================================================
#>                                iNZight Inference using Normal Theory
#> ----------------------------------------------------------------------------------------------------
#>    Primary variable of interest: Sepal.Length (numeric)
#>              Secondary variable: Species (categorical)
#>                                  
#>    Total number of observations: 150
#> ====================================================================================================
#> 
#> Inference of Sepal.Length by Species:
#> -------------------------------------
#> 
#> Group Means with 95% Confidence Intervals
#> 
#>                 Lower   Estimate   Upper
#>        setosa   4.906      5.006   5.106
#>    versicolor   5.789      5.936   6.083
#>     virginica   6.407      6.588   6.769
#> 
#> One-way Analysis of Variance (ANOVA F-test)
#> 
#>    F = 119.26, df = 2 and 147, p-value < 2.22e-16
#> 
#>           Null Hypothesis: true group means are all equal
#>    Alternative Hypothesis: true group means are not all equal
#> 
#> 
#> ### Difference in mean Sepal.Length between Species groups
#>     (col group - row group)
#> 
#> Estimates
#> 
#>                 setosa   versicolor
#>    versicolor   -0.930             
#>     virginica   -1.582       -0.652
#> 
#> 95% Confidence Intervals (adjusted for multiple comparisons)
#> 
#>                  setosa   versicolor
#>    versicolor   -1.1738             
#>                 -0.6862             
#>     virginica   -1.8258      -0.8958
#>                 -1.3382      -0.4082
#> 
#> P-values
#> 
#>                 setosa   versicolor
#>    versicolor        0             
#>     virginica        0            0
#> 
#> 
#> ====================================================================================================
```

These use the same notation as the plot, and are intended to be used
alongside.

## Features

There are many, many features available in `iNZightPlots`, all of which
are added simply through additional arguments to `inzplot()`. For a full
list, check `?iNZightPlot` (this is the base function for which
`inzplot()` is a convenience wrapper) and `?inzpar`.

### Colour and size

Probably the most common features to add to a graph, we have the
`sizeby` and `colby` arguments which map an additional variable to size
and colour, respectively:

``` r
inzplot(Sepal.Length ~ Sepal.Width, data = iris, colby = Species)
```

<img src="man/figures/README-colours-1.png" width="50%" style="display: block; margin: auto;" />

### Inference lines

Usually you want to know if what you see has any statistical
significance, and `iNZightPlots` makes this easy. We also provide
**comparison intervals** which can be, very simply, interpreted as
“overlapping comparison intervals indicate no significant difference
between groups”.

``` r
# Add confidence intervals
inzplot(Sepal.Length ~ Species, data = iris,
  inference.type = c("conf", "comp"), # red and black, respectively
  inference.par = "mean"
)
```

<img src="man/figures/README-demo-inf.png" width="50%" />

For scatter plots, inference uses linear trends. You can add a bootstrap
sample of trend estimates using the `bs.inference` argument:

``` r
inzplot(Sepal.Length ~ Sepal.Width, data = iris,
  trend = "linear",
  bs.inference = TRUE
)
```

<img src="man/figures/README-trends-2.png" width="50%" style="display: block; margin: auto;" />


NEWS.Md
# iNZightPlots 2.14

- add `ci.width` argument to `inzpar()` for plots and inferences for their confidence intervals
- suppress warnings from `RColorBrewer::brewer.pal` if `n <= 2`
- fix bug in ggplot2 plots where discrete axis ordering was different to a regular iNZightPlot (#270, @daniel-barnett)
- add `table.direction` argument to `inzsummary()` to display tables horizontally (default) or vertically
- reorganise output for inference (differences, CIs, and p-values) into a single table, rather than 2 or 3 matrices
- survey anova inference now provides CIs and adjusted p-values using 'emmeans' package
- one- and two-way tables' inference displays a single table, and uses correct method for calculating CIs for differences in proportions in a survey design
- clarify wording for two-way difference tables
- allow specification of random rounding and suppression rules
  - one- and two-way tables
  - suppress values based on small counts/large RSE, and other privacy controls

## iNZightPlots 2.13.5

- patch issue with summary/inference if the graphics device has minimal/no dimensions

## iNZightPlots 2.13.4

- fix bug in `inzsummary` and `inzinference` for survey regression output (response/predictor wrong way around)
- disable plots for surveys with x and/or y with only 1 level (#258, @tmelliott)
- fix handling of axis transformations for survey data (#227, @tmelliott)

## iNZightPlots 2.13.3

- update use of `svyquantile` for changes in 'survey' 4.1

## iNZightPlots 2.13.2

- fix CRAN issues on R-devel (windows), and M1 mac (all related to tests)

## iNZightPlots 2.13.1

- fix issue in exporting as HTML with local assets (@tmelliott, #252)

# iNZightPlots 2.13

- replace `inzsummary` and `inzinference` with S3 methods
- only show survey information (exclude the full call) in summary/inference
- fix bug where plots failed if there we no non-missing x/y values in a subset
- fix issue introduced by change in 'readr' package (uses 'hms' instead of 'time')
- fix y-axis rendering of large values (now accounts for width of tick labels)

## iNZightPlots 2.12.6

- fix wording in inference information output ('mean' -> 'estimate') (@tmelliott, #242)

## iNZightPlots 2.12.5

- add Mantel-Haenszel tests when `g1` is specified in `getPlotSummary`
- fix ordering of anova output (col - row instead of row - col)

## iNZightPlots 2.12.4

- pass `na.rm = TRUE` to survey methods so summaries and inference show the necessary information
- fix bug in `inzplot()` causing axes to be the wrong way around in hex plots for surveys
- `smooth` is respected for survey data (passed to `svysmooth`)

## iNZightPlots 2.12.3

- fix bug in survey inference where levels were the wrong way around (A - B instead of B - A)

## iNZightPlots 2.12.2

- fix bug in plots with axis range of zero (i.e., all x/y-values equal)

## iNZightPlots 2.12.1

- fix bug so mean indicator on dot plots works with surveys

# iNZightPlots 2.12

This release mainly consists of compatibility changes to accompany 'iNZight' 4.0.

### Breaking changes

- New `inzplot`, `inzsummary`, and `inzinference` replace `iNZPlot`, `iNZSummary`, and `iNZInference`, providing a better interface to the package which better first the R ecosystem.

### Minor changes

- refactors exportHTML() and exportSVG() to writing file to (by default) a temporary directory, and without changing directories in the process
- handle `col.emph`, `col.emphn`, and (by default) plot emphasized points on top
- allow `col.fun` to be a character palette name
- when dot plots show inference, override `boxplot` and `mean_indicator` (depending on specification of `inference.par`)
- add normal inference to scatter plot trend lines
- (survey) pop size estimates shown even if weights seem odd

### Bug fixes

- fix handling of env argument
- handle "id" value for `locate` argument
- mean indicator scales with overall plot scale, not dot point size
- fix handling of `local` argument (for exportHTML())

---

## iNZightPlots 2.11.6

- fix bug resulting in an invalid p-value being displayed for some two sided tests

## iNZightPlots 2.11.5

- provide design effects for survey summaries
- fix bug where regions cannot be selected in interactive sparkline plots due to id issue
- allow `main` to include templates for `%subet%` and `%sizeby%`
- fix y-axis to correctly display percentages (as percentages)
- rename `const_palette_names()` to `cont_palette_names()` (originally misspelled)
- fix bug in checking design call is not `as.svrepdeign`

## iNZightPlots 2.11.4

- fix bug that causes incorrect interactive map labels due to multipolygon regions

## iNZightPlots 2.11.3

- fix bug in exporting of a single numeric variable

## iNZightPlots 2.11.2

- fix a bug in the arguments of the getInfo() method causing exportHTML() to fail

## iNZightPlots 2.11.1

- fix formula interface to handle `colby`, etc.
- fix FT plots for numeric/factor so either order displays the same plot (to match base iNZightPlots graphs)
- Move Financial Times graph package dependencies to Suggests (from Imports)
- Specify `stringsAsFactors = TRUE` for upcoming R 4.0.0

# iNZightPlots 2.11

- Add formula interface function, `iNZPlot`, for `iNZightPlot`, along with similar wrappers `iNZSummary` and `iNZInference` for summary and inferential information, respectively.
- Get Summary for trends no longer includes superfluous "+" in front of negative coefficients
- Make subsetting/faceting for ggplots more in line with base iNZightPlots
- Fix bug in summary output where "Population" was being displayed for non-survey data
- Fix bug in axis labelling of large values where scientific notation used inconsistently

# iNZightPlots 2.10.4

**Release date**: 11 November 2019

- Add `use.plotly` attribute to ggplots
- New method of plotting barcode plots using `geom_spoke`
- Add cutpoint option for diverging stacked bar charts
- Sortable plots now can be sorted in either ascending or descending
- Gridplots now ignore background colours

R/inzpar.R
#' Plotting parameters for iNZight Plots
#'
#' A whole suite of parameters that can be used to fine-tune plots obtained from the
#' \code{iNZightPlot} function.
#' The parameters include both plot type, style, and appearance.
#'
#' @details
#'
#' \describe{
#' \item{'pch'}{the plotting symbol to be used; default is `21` (circle with fill)}
#' \item{'col.pt'}{the colour of points. this can either be a single value, or a vector of
#' colours if \code{colby} is specified}
#' \item{'col.fun'}{a function to use for colouring points, etc., or the name of a palette,
#'  see \code{inzpalette}
#' }
#' \item{'col.emph', 'col.emphn'}{emphasize the chosen level of a colour by variable. For numeric
#'  colour by, \code{col.emphn} specifies the number of quantiles to use.
#' }
#' \item{'emph.on.top'}{if \code{TRUE}, emphasised points will be positioned on top}
#' \item{'col.default'}{the default colour functions, containing a list with entries for 'cat' and 'cont' variables}
#' \item{'col.missing'}{the colour for missing values; default is a light grey}
#' \item{'reverse.palette'}{logical, if \code{TRUE} the palette will be reversed}
#' \item{'col.method'}{the method to use for colouring by a variable, one of 'linear' or 'rank'}
#' \item{'cex'}{the overall scaling for the entire plot; values less than 1 will make the text and points
#' smaller, while values larger than 1 will magnify everything}
#' \item{'cex.pt'}{the scaling value for points}
#' \item{'cex.dotpt'}{
#'      the scaling value for points in a dotplot. Note, this is not multiplicative with
#'      \code{'cex.pt'}
#' }
#' \item{'cex.lab'}{the scaling value for the plot labels}
#' \item{'cex.axis'}{the scaling value for the axis labels}
#' \item{'cex.main'}{the scaling value for the main plot title}
#' \item{'cex.text'}{the scaling value for text on the plot}
#' \item{'resize.method'}{one of 'proportional' (default) or 'emphasize'}
#' \item{'alpha'}{transparency setting for points; default is 1, 0 is fully transparent}
#' \item{'bg'}{the background colour for the plot}
#' \item{'grid.lines'}{logical to control drawing of axis grid lines}
#' \item{'col.grid'}{if 'grid.lines' is \code{TRUE}, this controls the colour of them.
#'      The default is 'default', which will choose a colour based on the value of 'bg')
#' }
#' \item{'fill.pt'}{the fill colour for points; default is \code{"transparent"}}
#' \item{'lwd'}{the line width of lines (for joining points)}
#' \item{'lty'}{the line type of lines (for joining points)}
#' \item{'lwd.pt'}{the line width used for points; default is 2}
#' \item{'col.line'}{the colour of lines used to join points}
#' \item{'col.sub'}{vector of up to two colours for the background of subplot labels.
#'      If only one specified, it is used for both.
#' }
#' \item{'locate.col.def'}{the default colour for locating points}
#' \item{'highlight.col'}{colour to use for highlighting points}
#' \item{'jitter'}{the axes to add jitter to. Takes values \code{"x"}, \code{"y"},
#'      or \code{"xy"} (default is en empty string, \code{""})}
#' \item{'rugs'}{the axes to add rugs to. Takes same values as \code{jitter}}
#' \item{'trend'}{a vector containing the trend lines to add to the plot. Possible values
#'      are \code{c("linear", "quadratic", "cubic")}}
#' \item{'smooth'}{the smoothing (lowess) for the points. Takes a value between 0 and 1
#'      (the default, 0, draws no smoother)}
#' \item{'smoothby.lty'}{the line type used for smoothers if \code{trend.by = TRUE}}
#' \item{'quant.smooth'}{if quantile smoothers are desired, they can be specified here as either
#'      the quantiles to smooth over (e.g., \code{c(0.25, 0.5, 0.75)}), or \code{"default"}, which
#'      uses the sample size to decide on an appropriate set of quantile smoothers}
#' \item{'LOE'}{logical, if \code{TRUE}, then a 1-1 line of equality is drawn}
#' \item{'join'}{logical, if \code{TRUE}, then points are joined by lines}
#' \item{'lines.by}{logical, if \code{join = TRUE} and \code{colby} is specified, points are joined
#'      by the specified variable}
#' \item{'col.trend'}{a named list of colours to be used for drawing the lines. The default is
#'      \code{list(linear = "blue", quadratic = "red", cubic = "green4")}}
#' \item{'lty.trend'}{a named list of line types for various types of trend lines. The default is
#'      \code{list(linear = 1, quadratic = 2, cubic = 3)}}
#' \item{'trend.by'}{logical, if \code{TRUE}, then trend lines are drawn separately for
#'      each group specified by \code{colby}}
#' \item{'trend.parallel'}{logical, if \code{TRUE}, the trend lines by group are given the same slope;
#'      otherwise they are fit independently}
#' \item{'col.smooth'}{the colour of the smoother}
#' \item{'col.LOE'}{the colour of the line of equality}
#' \item{'lty.LOE'}{the line type of the line of equality}
#' \item{'boxplot'}{logical, if \code{TRUE}, a boxplot is drawn with dotplots and histograms}
#' \item{'box.lwd', 'box.col', 'box.fill'}{the line width, colour, and fill colour for
#'      the box plot drawn}
#' \item{'bar.lwd', 'bar.col', 'bar.fill'}{the line width, colour, and fill colour of bars in a bar plot}
#' \item{'bar.counts'}{logical, if \code{TRUE} bar graphs will display counts instead of percentages (the default)}
#' \item{'full.height'}{may no longer be necessary ...}
#' \item{'inf.lwd.comp', 'inf.lwd.conf'}{the line width of comparison and confidence intervals, respectively}
#' \item{'inf.col.comp', 'inf.col.conf'}{the colour of comparison and confidence intervals, respectively.
#'      These take a length 2 vector, where the first element is used for normal inference, while the second
#'      is used for bootstrap intervals}
#' \item{'inference.type'}{the type of inference added to the plot. Possible values
#'      are \code{c("comp", "conf")}}
#' \item{'inference.par'}{the parameter which we obtain intervals for. For a dotplot or histogram,
#'      this can be either \code{"mean"} or \code{"median"}; for bar plots it can be "proportion"}
#' \item{'ci.width'}{the width of confidence intervals, default 0.95 for a 95\% confidence interval}
#' \item{'bs.inference'}{logical, if \code{TRUE}, then nonparametric bootstrap simulation is used
#'      to obtain the intervals}
#' \item{'inference.method'}{one of 'normal', 'bootstrap', or 'bayesian'}
#' \item{'min.count'}{the min count for barplots inference; counts less than this are ignored}
#' \item{'n.boot'}{the number of bootstrap simulations to perform}
#' \item{'large.sample.size'}{sample sizes over this value will use a large-sample plot variant
#'      (i.e., scatter plots will become hex plots, dot plots become histograms)}
#' \item{'largesample'}{logical, if \code{TRUE}, then the large-sample plot variance is used}
#' \item{'scatter.grid.bins'}{the number, N, of bins to use for the scatter-grid plot,
#'      producing an N x N matrix}
#' \item{'hex.bins'}{the number of bins to use for hexagonal binning}
#' \item{'hex.style'}{the style of the hexagons, one of "size" or "alpha"}
#' \item{'hex.diffuse'}{logical, Pass on rounding error to nearest not-yet-drawn hexes so that
#'      rare classes get represented}
#' \item{'hist.bins'}{the number of bins to use for the histogram
#'      (The default \code{NULL} uses point size to approximate dot plot)}
#' \item{'quant.cutoff'}{if \code{quant.smooth = "default"}, these sample size values are used
#'      to determine which quantiles are drawn}
#' \item{'plottype'}{used to override the default plot type. Possible values, depending on data type,
#'      include \code{c("scatter"|"grid"|"hex"|"dot"|"hist")}}
#' \item{'matchplots'}{logical, if \code{TRUE}, then the type of plot is kept consistent between different
#'      subsets}
#' \item{'match.limits'}{a vector of two values used to decide whether to use all small-sample or all
#'      large-sample plots}
#' \item{'xlim'}{a vector defining the x axis limits (default NULL will use the data)}
#' \item{'ylim'}{a vector defining the y axis limits (default NULL will use the data)}
#' \item{'transform'}{a list of variable transformations (e.g., list(x = 'log'))}
#' \item{'plot.features'}{a list containing any additional features for new plots (e.g., maptype)}
#' \item{'round'}{integer specifying optional rounding of numerical output, default NA (ignored)}
#' \item{'round_percent'}{integer specifying rounding for percentages (default 2)}
#' \item{'signif'}{integer specifying number of significant figured in numeric output (default 2). Ignored if \code{round} is not NA.}
#' }
#'
#' @title iNZight Plotting Parameters
#' @param ... If arguments are supplied, then these values are set. If left empty, then
#' @param .viridis checks if the viridis package is installed; or can be turned off
#' the default list is returned.
#' @return an object of class \code{inzpar.list}
#' @export
#' @examples
#' # arguments can be passed directly to \code{iNZightPlot}
#' iNZightPlot(Sepal.Length, data = iris, col.pt = "red",
#'     box.col = "blue", box.fill = "green")
#'
#' # or stored and passed to it (only pars relevant to the current
#' # plot are used)
#' mypar <- inzpar(col.pt = "red", box.col = "blue", box.fill = "green",
#'     trend = "linear", trend.by = TRUE)
#' inzplot(Sepal.Length ~ Species, data = iris, inzpar = mypar)
#' iNZightPlot(Sepal.Length, Sepal.Width, data = iris, inzpar = mypar,
#'     colby = Species)
inzpar <- function(...,
                   .viridis = requireNamespace("viridis", quietly = TRUE)) {
    dots <- list(...)

    ip <- list(
        pch            = 21L,
        col.pt         = "grey50",
        col.fun        = NULL,
        col.emph       = 0L,
        col.emphn      = 4L,
        emph.on.top    = TRUE,
        col.default    =
            list(
                cat  =
                    if (.viridis) {
                        function(n)
                            if (n > 10) viridis::viridis(n)
                            else c(
                                "#E69F00", "#56AAE9", "#D55E00", "#0072B2",
                                "#F0D705", "#ADD9FF", "#9BCD9B", "#CC79A7",
                                "#68468C", "#8B0000"
                            )[1:n]
                    } else {
                        function(n)
                            if (n > 10) hcl( (1:n) / n * 360, c = 80, l = 50)
                            else c(
                                "#E69F00", "#56AAE9", "#D55E00", "#0072B2",
                                "#F0D705", "#ADD9FF", "#9BCD9B", "#CC79A7",
                                "#68468C", "#8B0000"
                            )[1:n]
                    },
                cont =
                    if (.viridis) viridis::viridis
                    else function(n)
                        hcl( (1:n) / n * 320 + 60, c = 100, l = 50)
            ),
        col.missing    = "#cccccc",
        reverse.palette = FALSE,
        col.method     = "linear",
        cex            = 1,
        cex.pt         = 0.8,
        cex.dotpt      = 0.5,
        cex.lab        = 1,
        cex.axis       = 0.8,
        cex.main       = 1.2,
        cex.text       = 1,
        resize.method  = "proportional",
        alpha          = 1,
        bg             = "#eeeeee",
        grid.lines     = TRUE,
        col.grid       = "default",
        fill.pt        = "transparent",
        lwd            = 1,
        lty            = 1,
        lwd.pt         = 2,
        col.line       = "blue",
        col.sub        = c("#cccccc", "#444444"),
        locate.col.def = "red",
        highlight.col  = "shade",
        jitter         = "",
        rugs           = "",
        trend          = NULL,
        smooth         = 0,
        smoothby.lty   = 4,  # MUST be numeric
        quant.smooth   = NULL,
        LOE            = FALSE,
        join           = FALSE,
        lines.by       = TRUE,
        col.trend =
            list(
                linear = "blue",
                quadratic = "red",
                cubic = "green4"
            ),
        lty.trend =
            list(
                linear = 1,
                quadratic = 2,
                cubic = 3
            ),
        trend.by       = FALSE,
        trend.parallel = TRUE,
        col.smooth     = c("magenta"),
        col.LOE        = "black",
        lty.LOE        = 2,
        mean_indicator = FALSE,
        boxplot        = TRUE,
        box.lwd        = c(2, 0.7),
        box.col        = "black",
        box.fill       = "grey90",
        bar.lwd        = 1,
        bar.col        = "black",
        bar.fill       = "darkgreen",
        bar.counts     = FALSE,
        full.height    = FALSE,
        inf.lwd.comp   = 4,
        inf.lwd.conf   = 2,
        inf.col.comp   = c("black", "green4"),
        inf.col.conf   = c("red", "orange"),
        inference.type = NULL,
        inference.par  = NULL,
        ci.width       = 0.95,
        bs.inference   = FALSE,
        inference.method = "normal",
        min.count      = 5,
        n.boot         = 1500,
        large.sample.size = 5001,
        largesample    = NULL,
        scatter.grid.bins = 50,
        hex.bins       = 20,
        hex.style      = "size",
        hex.diffuse    = TRUE,
        hist.bins      = NULL,
        quant.cutoff   = c(200, 1000),
        plottype       = "default",
        matchplots     = TRUE,
        match.limits   = c(500, 10000),
        internal.labels = TRUE,
        xlim           = NULL,
        ylim           = NULL,
        transform      = list(),
        plot.features  = list(),
        epi.out        = FALSE,
        round          = NA_integer_,
        round_percent  = 2L,
        signif         = 4L
    )

    # update any user has specified
    if (length(dots) > 0) {
        for (i in names(dots))
            ip[[i]] <- dots[[i]]
    }

    class(ip) <- "inzpar.list"
    ip
}

# Used internally
valid_par <- function(par, plot_type, what = c("plot", "summary", "inference")) {
    what <- substr(match.arg(what), 1, 1)
    res <- rep(TRUE, length(par))
    if (! plot_type %in% colnames(plot_types) ) {
        warning("Invalid or unknown plot type")
        return(res)
    }
    v <- par %in% rownames(plot_types)
    res[v] <- grepl(what, plot_types[par[v], plot_type])
    res
}

gg_defaults <- list(
    adjust = 1,
    alpha_densitygroup = 0.6,
    gg_lwd = 1,
    gg_size = 6,
    ordered = FALSE,
    gg_barSize = 16,
    gg_bins = 30,
    gg_height = 0.5,
    gg_width = 1,
    gg_perN = NULL,
    gg_swarmwidth = 0.4,
    gg_method = "quasirandom",
    gg_cutpoint = "Default",
    gg_theme = "grey",
    fill_colour = "",
    rotation = FALSE,
    rotate_labels = list(x = FALSE, y = FALSE),
    caption = "",
    desc = FALSE,
    palette = "default"
)
